### ISSUE TITLE
OpenVINO not getting installed via Dockerfile.

### ISSUE THREAD

**Issue by synchronizing** [opencv/gst-video-analytics#68]
It seems that OpenVINO is not getting installed in the container via both the [binaries.Dockerfile](https://github.com/opencv/gst-video-analytics/blob/master/docker/binaries.Dockerfile) and [Dockerfile](https://github.com/opencv/gst-video-analytics/blob/master/docker/Dockerfile) provided.

For `binaries.Dockerfile` specifically, In the following line:

https://github.com/opencv/gst-video-analytics/blob/a1f5efb660d3916d63f2a26d23e246f8d8651634/docker/binaries.Dockerfile#L557

I replaced it with:

```Dockerfile
RUN wget http://registrationcenter-download.intel.com/akdlm/irc_nas/16612/l_openvino_toolkit_p_${OpenVINO_VERSION}.tgz
```

Which gave me successful execution, and thereafter a successful container run (plus, clear output for `l_openvino_toolkit_p_2020.2.120/install.sh` being ran). However, when exploring the file system (and going down to the tutorial for ["Acquire models and videos"](https://github.com/opencv/gst-video-analytics/wiki/Install-Guide#3-acquire-models-and-videos-) the following instructions:

```shell
source /opt/intel/openvino/bin/setupvars.sh
source ~/gva/gst-video-analytics/scripts/setup_env.sh

# Download models listed in models.lst into ~/intel/dl_streamer/models folder
cd ~/gva/gst-video-analytics/samples
./download_models.sh
```

Does not work because `/opt/intel/openvino` is not a path in the container.

```shell
root@4aac8f4d2944:~# ls /opt/intel/
dldt  mediasdk
```

This is the same when building with `opensource` via the [Dockerfile](https://github.com/opencv/gst-video-analytics/blob/master/docker/Dockerfile) provided:

```shell
root@bf6b3402dee1:~# ls /opt/intel/
dldt  mediasdk
```

Any ideas what this could be? I've been able to create a custom docker file that installs all the required dependencies (`gstreamer`, `dl-streamer`, and `OpenVINO`), but when running `gstreamer` with the `dl-streamer` plugins + OpenVino models I receive multiple warnings and the output video is the same as the input. My current assumption is that `dl-streamer` does not support latest versions of `gstreamer`+plugins... but I digress. Currently exploring the above route again (using the provided Dockerfiles in the repo), but if anything, I'll open a new issue discussing revamping the Dockerfile in this repo.

**Comment by brmarkus** [comment]
In the past the URL varied between the releases plus allowing the user to flexibly use other downloaded variants (e.g. alpha-releases, internal releases, pre-releases, etc.), therefor it wasn't downloaded while building the Docker container image.

In the meantime GVA is already part of the ("native") OpenVINO installation, and is now called "DL-streamer".

**Comment by umed** [comment]
hey @synchronizing,

> source /opt/intel/openvino/bin/setupvars.sh
source ~/gva/gst-video-analytics/scripts/setup_env.sh

You can skip that part for `opensource` build. They are only required in local installation or in `binaries` build. That part of instruction will be fixed soon.

However, by output of `ls /opt/intel` in both cases it looks like you are in the container based on `opensource` build from this [docker/Dockerfile](https://github.com/opencv/gst-video-analytics/blob/master/docker/Dockerfile). Could you verify this?

**Comment by umed** [comment]
I rechecked [binaries.Dockerfile](https://github.com/opencv/gst-video-analytics/blob/master/docker/binaries.Dockerfile). Seems like you don't need to run these steps at all for both `opensource` or `binaries` build:

>source /opt/intel/openvino/bin/setupvars.sh
source ~/gva/gst-video-analytics/scripts/setup_env.sh

Also, output of `ls /opt/intel` should look the same. Seems like we a bit changed our `binaries.Dockerfile` and I missed that 😃

**Comment by synchronizing** [comment]
> In the meantime GVA is already part of the ("native") OpenVINO installation, and is now called "DL-streamer". 
> @brmarkus 

Those were my thoughts at first. The first thing I did was try to use the Dockerfile provided by the OpenVINO account on DockerHub ([here](https://hub.docker.com/u/openvino)). The Dockerfile allows me to install DL Streamer via a `.sh` file provided, but it will not install `gstreamer`. The documentations on the OpenVINO website does not make it obvious what to do.

> However, by output of ls /opt/intel in both cases it looks like you are in the container based on opensource build from this docker/Dockerfile. Could you verify this?
> @umed 

I tried with _both_ the `opensource`, and the binaries installation, but I see you clarified that it shouldn't be needed to `source` the first variable. Let me give it a try, and I'll come back here some point today.

**Comment by easychizy** [comment]
@synchronizing,

Based on my understanding, the script runs model downloader "$DOWNLOADER" of OpenVINO to download the model files from [Open Model Zoo site](https://github.com/opencv/open_model_zoo). You don't have to port from host machine.

Does this answer your question?

Mark

**Comment by synchronizing** [comment]
> @easychizy 
> Based on my understanding, the script runs model downloader "$DOWNLOADER" of OpenVINO to download the model files from [Open Model Zoo site](https://github.com/opencv/open_model_zoo). You don't have to port from host machine.

Hey Mark. `$DOWNLOADER` is a reference to a non-existing path:

```shell
DOWNLOADER=$INTEL_OPENVINO_DIR/deployment_tools/open_model_zoo/tools/downloader/downloader.py
```

The `$INTEL_OPENVINO_DIR` is set by the start up script (if I remember correctly), and that is `/opt/intel/openvino`, which is missing from the container.

**Comment by synchronizing** [comment]
So I tried what I had mentioned earlier:

> I'm assuming that we are to port over models from the host machine to the Docker container on the docker run stage.

I downloaded the sample models and built the `package` version  (as described [here](https://github.com/opencv/gst-video-analytics/issues/68#issue-633686429)) with a `/root/files/` folder containing the sample models + sample video:

```
root@c2e63997bb5a:~# ls files/
emotions-recognition-retail-0003.json  emotions-recognition-retail-0003.xml  face-detection-adas-0001.xml  video.mp4
```

On running the following command:

```shell
gst-launch-1.0 \
filesrc location=/root/files/video.mp4 ! \
decodebin ! \
gvadetect \
  model=/root/files/face-detection-adas-0001.xml ! \
gvaclassify \
   model=/root/files/emotions-recognition-retail-0003.xml \
   model-proc=/root/files/emotions-recognition-retail-0003.json ! \
gvawatermark ! \
filesink location=/root/output.mp4
```

I receive the following output from gstreamer:

```
Setting pipeline to PAUSED ...
Pipeline is PREROLLING ...
Redistribute latency...
ERROR: from element /GstPipeline:pipeline0/GstGvaDetect:gvadetect0: base_inference plugin intitialization failed
Additional debug info:
/root/gst-video-analytics/gst/inference_elements/base/inference_singleton.cpp(118): acquire_inference_instance (): /GstPipeline:pipeline0/GstGvaDetect:gvadetect0:

Failed to construct OpenVINOImageInference
 Failed to import pre-compiled model
  [NOT_IMPLEMENTED]

ERROR: pipeline doesn't want to preroll.
Setting pipeline to NULL ...
Freeing pipeline ...
```

The `NOT_IMPLEMENTED` makes me assume this is due to the original reason for opening this issue, which is my belief OpenVINO never gets successfully installed. Unsure if this is the case, but any guidance would be wonderful!

**Comment by kunalg06** [comment]
Not able to cover the [Acquire models and videos](https://github.com/opencv/gst-video-analytics/wiki/Install-Guide#3-acquire-models-and-videos-) after installing via docker method. Can someone guide !!!

**Comment by brmarkus** [comment]
Please undo all changes in scripts and Dockerfile first :-) (because using Ubuntu 16 or Ubuntu 18 with GNOME deskopt/window-manager on an Intel Core-i9 is working for me).
I'm using "l_openvino_toolkit_p_2020.2.120.tgz", downloaded separately and copied it to the HOST-machine.

As I'm behind a corporate network proxy I needed to set the proxy ENV variables (http_proxy, https_proxy, HTTP_PROXY, HTTPS_PROXY), plus setting DISPLAY and calling "xhost +local:root".
I just used any MP4 video-container with embedded h.264 elementary stream. (as my computer doesn't have an Intel GPU I don't have HW-accelerated video-decoding via MediaSDK).

The needed models are downloaded successfully on my HOST-machine - and then using Docker "volume-mounting" when calling "docker run":

    -v /My/Path/gva/data/models/intel:/root/intel_models:ro \
    -v /My/Path/gva/data/models/common:/root/common_models:ro \
    -e MODELS_PATH=/root/intel_models:/root/common_models \

**Comment by kunalg06** [comment]
> Please undo all changes in scripts and Dockerfile first :-) (because using Ubuntu 16 or Ubuntu 18 with GNOME deskopt/window-manager on an Intel Core-i9 is working for me).
> I'm using "l_openvino_toolkit_p_2020.2.120.tgz", downloaded separately and copied it to the HOST-machine.
> 
> As I'm behind a corporate network proxy I needed to set the proxy ENV variables (http_proxy, https_proxy, HTTP_PROXY, HTTPS_PROXY), plus setting DISPLAY and calling "xhost +local:root".
> I just used any MP4 video-container with embedded h.264 elementary stream. (as my computer doesn't have an Intel GPU I don't have HW-accelerated video-decoding via MediaSDK).
> 
> The needed models are downloaded successfully on my HOST-machine - and then using Docker "volume-mounting" when calling "docker run":
> 
> ```
> -v /My/Path/gva/data/models/intel:/root/intel_models:ro \
> -v /My/Path/gva/data/models/common:/root/common_models:ro \
> -e MODELS_PATH=/root/intel_models:/root/common_models \
> ```

and how to download models to running docker? 
# Download models listed in models.lst into ~/intel/dl_streamer/models folder
cd ~/gva/gst-video-analytics/samples
./download_models.sh
after this Im getting 
"Downloading models to folder /root/intel_models:/root/common_models
python3: can't open file '/deployment_tools/open_model_zoo/tools/downloader/downloader.py': [Errno 2] No such file or directory"

**Comment by brmarkus** [comment]
I have installed OpenVINO in my HOST machine (i.e. outside of the Docker container).
By calling "source /opt/intel/openvino/bin/setupvars.sh" the OpenVINO env variables and paths are exposed, e.g. INTEL_OPENVINO_DIR.
When downloading the models on HOST (i.e. outside of the Docker container) then set the env variable e.g. "export MODEL_PATH=/My/Path/gva/data/models/", (and "make -p /My/Path/gva/data/models/") the downloaded models then will be in subfolders "intel" and "common".

**Comment by umed** [comment]
>Let me use this opportunity to ask you as well @umed -- I hope you don't mind.
>
> I've been working on porting over the project to a new Dockerfile for easier deployment, and I'm running into some issues where the video file passes through my CLI gstream+DL Streamer pipeline successfully (though with some warnings with debug level 4), but the expected video output does not contain drawn boxes (facial reconition + emotion detection, as the example in the wiki here). I was wondering if DL Streamer requires specific version dependencies for gstreamer and its good, bad, and ugly gstreamer plugins. A clarification here would be wonderful.

@synchronizing, short answer is our inference elements does not draw boxes, they attaches meta data, but we have `gvawatermark` (should be placed after inference elements) that draws meta on video frames. If that does not solve your problem please create separate issue and we can continue discussion there.

**Comment by synchronizing** [comment]
> @brmarkus 
> I have installed OpenVINO in my HOST machine (i.e. outside of the Docker container).

Would you know if this is a _must_ for it to work? My purpose with the Docker container is to eventually deploy it as an API pipeline, so having OpenVINO installed in the host machine isn't much of an option. 

Thank you for taking your time to chime in what worked for you. Honeslty appreciated. 

> @umed 
> short answer is our inference elements does not draw boxes, they attaches meta data, but we have `gvawatermark` (should be placed after inference elements) that draws meta on video frames. If that does not solve your problem please create separate issue and we can continue discussion there.

I'll open a new issue from here then. I'm indeed using `gvawatermark` ([ref](https://github.com/opencv/gst-video-analytics/issues/68#issuecomment-641033529)), but for whatever reason still getting runtime errors when trying to pipe videos through it. With @brmarkus insights, I think I know why now. Thank you for taking the time to answer, and my apologies for the thread growing so large. 

In short, the key thing I believe I've gathered from the discussion above is that OpenVINO must be installed in the host machine, and DL Streamer may be installed via Docker. I'll be trying later today and tomorrow to setup a new Dockerfile to install all dependencies in a single container.

**Comment by nnshah1** [comment]
@synchronizing Which openvino docker image did you try?

Did you use the data-dev version: https://hub.docker.com/r/openvino/ubuntu18_data_dev 

This should have all dependencies for GStreamer and DL Streamer - Can you confirm?

**Comment by synchronizing** [comment]
> @nnshah1
> @synchronizing Which openvino docker image did you try?
> 
> Did you use the data-dev version: https://hub.docker.com/r/openvino/ubuntu18_data_dev
> 
> This should have all dependencies for GStreamer and DL Streamer - Can you confirm?

I've been using `ubuntu18_dev` this entire time. I see on the link you mentioned:

```
This ubuntu18_data_dev image includes the following data processing components unlike ubuntu18_dev image:

DL Streamer*
GStreamer
Speech Libraries and End-to-End Speech Demos
```

The docs in OpenVINO does not seem to have details on this, neither does the wiki here. Thank you for pointing this out -- I'll give it a try and report back in.

**Comment by kunalg06** [comment]
> I have installed OpenVINO in my HOST machine (i.e. outside of the Docker container).
> By calling "source /opt/intel/openvino/bin/setupvars.sh" the OpenVINO env variables and paths are exposed, e.g. INTEL_OPENVINO_DIR.
> When downloading the models on HOST (i.e. outside of the Docker container) then set the env variable e.g. "export MODEL_PATH=/My/Path/gva/data/models/", (and "make -p /My/Path/gva/data/models/") the downloaded models then will be in subfolders "intel" and "common".

I need ask one more thing, while installing opentoolkit on host machine, im not able to install Intel® Graphics Compute Runtime for OpenCL". And FYI im using ubuntu 16.04 there are no drivers for ubuntu16. if so can tell how to do it!! thank you in advance!!

**Comment by nnshah1** [comment]
> > @brmarkus
> > I have installed OpenVINO in my HOST machine (i.e. outside of the Docker container).
> 
> Would you know if this is a _must_ for it to work? My purpose with the Docker container is to eventually deploy it as an API pipeline, so having OpenVINO installed in the host machine isn't much of an option.
> 
> Thank you for taking your time to chime in what worked for you. Honeslty appreciated.
> 
> > @umed
> > short answer is our inference elements does not draw boxes, they attaches meta data, but we have `gvawatermark` (should be placed after inference elements) that draws meta on video frames. If that does not solve your problem please create separate issue and we can continue discussion there.
> 
> I'll open a new issue from here then. I'm indeed using `gvawatermark` ([ref](https://github.com/opencv/gst-video-analytics/issues/68#issuecomment-641033529)), but for whatever reason still getting runtime errors when trying to pipe videos through it, with an empty video output. With @brmarkus insights, I think I know why now. Thank you for taking the time to answer, and my apologies for the thread growing so large.
> 
> In short, the key thing I believe I've gathered from the discussion above and other opened issues is the following:
> 
> * OpenVINO must be installed in the host machine, and DL Streamer may be installed via Docker.
> * ([#56 (comment)](https://github.com/opencv/gst-video-analytics/issues/56#issuecomment-609760617)) DL Streamer has specific `gstreamer` dependencies.
> 
> I'll be trying later today and tomorrow to setup a new Dockerfile to install all dependencies in a single container. I'll leave the thread open for a day (if that sits well with the maintainers here) until I have a single Dockerfile working with OpenVINO + `gstreamer` + DL Stream.

@synchronizing  Another project that might be interesting / useful is our VA Serving which allows to easily create a pipeline service using DL Streamer. https://github.com/intel/video-analytics-serving, would be interested in any feedback you have on that as well.

**Comment by synchronizing** [comment]
> @nnshah1 
> @synchronizing  Another project that might be interesting / useful is our VA Serving which allows to easily create a pipeline service using DL Streamer. https://github.com/intel/video-analytics-serving, would be interested in any feedback you have on that as well. 

Oh man, that looks awesome, and exactly within the realm of tools I'm looking for. I'll definitely be check it out further, thank you.

**Comment by synchronizing** [comment]
Closing this. Ultimately, will be using the [`video-analytics-serving`](https://github.com/intel/video-analytics-serving) platform sent by @nnshah1. Thank you to everyone here who helped out!
